From 16ccd36b232b85961f4ed19df4b4a653401b9407 Mon Sep 17 00:00:00 2001
From: Harry Ciao <qingtao.cao@windriver.com>
Date: Thu, 5 Feb 2009 18:33:16 +0800
Subject: [PATCH] EDAC: CPC925 sensor delayed registration

Make CPC925 EDAC driver register its vMC sensors in delayed mode,
by making use of ipmi_smi_watcher structure.

Signed-off-by: Harry Ciao <qingtao.cao@windriver.com>
---
 drivers/edac/cpc925_edac.c |   54 +++++++++++++++++++++++++++++++++++++++-----
 1 files changed, 48 insertions(+), 6 deletions(-)

diff --git a/drivers/edac/cpc925_edac.c b/drivers/edac/cpc925_edac.c
index d330af7..56bbefc 100644
--- a/drivers/edac/cpc925_edac.c
+++ b/drivers/edac/cpc925_edac.c
@@ -43,6 +43,13 @@
 #define TO_VOID_FUNCTOR(x) (void(*)(struct edac_device_ctl_info *edac_dev))(x)
 
 #ifdef CONFIG_EDAC_VMC
+/*
+ * Meaning of vmc_intf_num:
+ * -1, vMC has not been loaded yet;
+ * >=0, vMC's interface number.
+ */
+static int vmc_intf_num = -1;
+
 static int init_vmc_sensors(void)
 {
 	int rc;
@@ -96,6 +103,9 @@ static int init_vmc_sensors(void)
 
 static void uninit_vmc_sensors(void)
 {
+	if (vmc_intf_num == -1)
+		return;
+
 	vmc_sensor_uninit(VMC_DRAM_UECC_FLT, cpc925_vmc_lun);
 	vmc_sensor_uninit(VMC_DRAM_CECC_FLT, cpc925_vmc_lun);
 
@@ -112,6 +122,37 @@ static void uninit_vmc_sensors(void)
 	vmc_release_lun(cpc925_vmc_lun);
 }
 
+/* Add sensors to vMC */
+static void edac_new_smi(int intf_num, struct device *device)
+{
+	if (strcmp(device->driver->name, VMC_DEVICE_NAME) == 0) {
+		vmc_intf_num = intf_num;
+		printk(KERN_INFO "EDAC: found vMC available at intf_num %d\n",
+			intf_num);
+		if (init_vmc_sensors())
+			printk(KERN_ERR "failed to initialize one or more "
+				"vMC sensors\n");
+	}
+}
+
+/*
+ * Note, vMC's SDR/SEL repositories would have been destroyed
+ * before this smi_gone() is called, so there is nothing left to do.
+ */
+static void edac_smi_gone(int intf_num)
+{
+	if (vmc_intf_num == intf_num)
+		vmc_intf_num = -1;
+
+	return;
+}
+
+static struct ipmi_smi_watcher vmc_watcher = {
+	.owner = THIS_MODULE,
+	.new_smi = edac_new_smi,
+	.smi_gone = edac_smi_gone
+};
+
 #endif /* CONFIG_EDAC_VMC */
 
 
@@ -546,11 +587,11 @@ static int __devinit cpc925_probe(struct platform_device *pdev)
 
 
 #ifdef CONFIG_EDAC_VMC
-	/* Initialize vMC Sensors */
-	if (init_vmc_sensors()) {
-		printk(KERN_ERR "failed to initialize one or more " \
-				"vMC sensors\n");
-	}
+        int ret = ipmi_smi_watcher_register(&vmc_watcher);
+        if (ret)
+                printk(KERN_ERR CPC925_EDAC_MOD_STR "can't register vMC's "
+                        "smi watcher!\n");
+
 #endif	/* CONFIG_EDAC_VMC */
 
 	/* get this far and it's successful */
@@ -581,7 +622,8 @@ static int cpc925_remove(struct platform_device *pdev)
 	edac_device_free_ctl_info(edac_dev);
 
 #ifdef CONFIG_EDAC_VMC
-	/* Uninitialize vMC sensors */
+	/* Unregister vMC's smi_watcher and cleanup vMC sensors */
+        ipmi_smi_watcher_unregister(&vmc_watcher);
 	uninit_vmc_sensors();
 #endif
 
-- 
1.6.0.3

